rust/datafusion/src/optimizer/constant_folding.rs

```diff
@@ -0,0 +1,620 @@
+            | LogicalPlan::Limit { .. }
+            | LogicalPlan::Repartition { .. }
+            | LogicalPlan::CreateExternalTable { .. }
+            | LogicalPlan::Extension { .. }
+            | LogicalPlan::Sort { .. }
+            | LogicalPlan::Explain { .. }
+            | LogicalPlan::Join { .. } => {
+                let expr = utils::expressions(plan);
+
+                // apply the optimization to all inputs of the plan
+                let inputs = utils::inputs(plan);
+                let new_inputs = inputs
+                    .iter()
+                    .map(|plan| self.optimize(plan))
+                    .collect::<Result<Vec<_>>>()?;
+
+                utils::from_plan(plan, &expr, &new_inputs)
+            }
+            LogicalPlan::TableScan { .. } | LogicalPlan::EmptyRelation { .. } => {
+                Ok(plan.clone())
+            }
+        }
+    }
+
+    fn name(&self) -> &str {
+        "constant_folding"
+    }
+}
+
+/// Recursively transverses the logical plan.
+fn optimize_expr(e: &Expr, schema: &DFSchemaRef) -> Result<Expr> {
+    Ok(match e {
+        Expr::BinaryExpr { left, op, right } => {
+            let left = optimize_expr(left, schema)?;
+            let right = optimize_expr(right, schema)?;
+            match op {
+                Operator::Eq => match (&left, &right) {
+                    (
+                        Expr::Literal(ScalarValue::Boolean(l)),
+                        Expr::Literal(ScalarValue::Boolean(r)),
+                    ) => Expr::Literal(ScalarValue::Boolean(Some(
+                        l.unwrap_or(false) == r.unwrap_or(false),
+                    ))),
+                    (Expr::Literal(ScalarValue::Boolean(b)), _)
+                        if right.get_type(schema)? == DataType::Boolean =>
+                    {
+                        match b {
+                            Some(true) => right,
+                            Some(false) | None => Expr::Not(Box::new(right)),
+                        }
+                    }
+                    (_, Expr::Literal(ScalarValue::Boolean(b)))
+                        if left.get_type(schema)? == DataType::Boolean =>
+                    {
+                        match b {
+                            Some(true) => left,
+                            Some(false) | None => Expr::Not(Box::new(left)),
+                        }
+                    }
+                    _ => Expr::BinaryExpr {
+                        left: Box::new(left),
+                        op: Operator::Eq,
+                        right: Box::new(right),
+                    },
+                },
+                Operator::NotEq => match (&left, &right) {
+                    (
+                        Expr::Literal(ScalarValue::Boolean(l)),
+                        Expr::Literal(ScalarValue::Boolean(r)),
+                    ) => Expr::Literal(ScalarValue::Boolean(Some(
+                        l.unwrap_or(false) != r.unwrap_or(false),
```

Comment:
What about if both operands are `NULL` ?

Reply:
It looks like that will result in `False`, I would expect `NULL::BOOL OR NULL::BOOL` to be `NULL`?
Probably no one will write this, but it could be the result of another optimization pass?

Reply:
yeah, any binary operation, other than null safe equality, that involves null should return null, i will updated the other non-literal branches to return null literal as well.